mock/src/network.rs

```diff
@@ -0,0 +1,183 @@
+use aleph_bft_types::{Network as NetworkT, NodeCount, NodeIndex, Recipient};
+use async_trait::async_trait;
+use futures::{
+    channel::mpsc::{unbounded, UnboundedReceiver, UnboundedSender},
+    Future, StreamExt,
+};
+use log::debug;
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+pub type NetworkReceiver<D> = UnboundedReceiver<(D, NodeIndex)>;
+pub type NetworkSender<D> = UnboundedSender<(D, NodeIndex)>;
+
+pub struct Network<D> {
+    rx: NetworkReceiver<D>,
+    tx: NetworkSender<D>,
+    peers: Vec<NodeIndex>,
+    index: NodeIndex,
+}
+
+impl<D> Network<D> {
+    pub fn new(
+        rx: NetworkReceiver<D>,
+        tx: NetworkSender<D>,
+        peers: Vec<NodeIndex>,
+        index: NodeIndex,
+    ) -> Self {
+        Network {
+            rx,
+            tx,
+            peers,
+            index,
+        }
+    }
+    pub fn index(&self) -> NodeIndex {
+        self.index
+    }
+}
+
+#[async_trait::async_trait]
+impl<D: Clone + Send> NetworkT<D> for Network<D> {
+    fn send(&self, data: D, recipient: Recipient) {
+        use Recipient::*;
+        match recipient {
+            Node(node) => self
+                .tx
+                .unbounded_send((data, node))
+                .expect("send on channel should work"),
+            Everyone => {
+                for peer in self.peers.iter() {
+                    if *peer != self.index {
+                        self.send(data.clone(), Node(*peer));
+                    }
+                }
+            }
+        }
+    }
+
+    async fn next_event(&mut self) -> Option<D> {
+        Some(self.rx.next().await?.0)
+    }
+}
+
+pub struct Peer<D> {
+    tx: NetworkSender<D>,
+    rx: NetworkReceiver<D>,
+}
+
+#[async_trait]
+pub trait NetworkHook<D>: Send {
+    /// This must complete during a single poll - the current implementation
+    /// of Router will panic if polling this method returns Poll::Pending.
+    async fn update_state(&mut self, data: &mut D, sender: NodeIndex, recipient: NodeIndex);
+}
+
+pub struct Router<D> {
+    peers: RefCell<HashMap<NodeIndex, Peer<D>>>,
+    peer_list: Vec<NodeIndex>,
+    hook_list: RefCell<Vec<Box<dyn NetworkHook<D>>>>,
+    reliability: f64, //a number in the range [0, 1], 1.0 means perfect reliability, 0.0 means no message gets through
```

Comment:
This comment might be more useful above `new`.

Reply:
moved